flow-tests/test-dev-mode/src/test/java/com/vaadin/flow/uitest/ui/UrlValidationIT.java

```
@@ -16,35 +16,66 @@
+
+import org.apache.commons.io.IOUtils;
 import org.junit.Assert;
 import org.junit.Test;
 
 public class UrlValidationIT extends ChromeBrowserTest {
 
     @Test
-    public void devModeUriValidation_uriWithDirectoryChange_statusForbidden() throws IOException {
+    public void devModeUriValidation_uriWithDirectoryChange_statusForbidden() throws IOException, InterruptedException {
         sendRequestAndValidateResponseStatusForbidden(
                 "/VAADIN/build/%252E%252E/");
     }
 
     @Test
-    public void staticResourceUriValidation_uriWithDirectoryChange_statusForbidden() throws IOException {
+    public void staticResourceUriValidation_uriWithDirectoryChange_statusForbidden() throws IOException, InterruptedException {
         sendRequestAndValidateResponseStatusForbidden(
                 "/VAADIN/build/%252E%252E/some-resource.css");
     }
 
-    private void sendRequestAndValidateResponseStatusForbidden(String pathToResource) throws IOException {
+    private void sendRequestAndValidateResponseStatusForbidden(String pathToResource) throws IOException, InterruptedException {
         final String urlString = getRootURL() + "/view" + pathToResource;
         URL url = new URL(urlString);
+
+        // not using chromedriver, so we need to look at the response body
+        // to know if the frontend is ready
+        waitForFrontendCompilation(url);
+
         HttpURLConnection connection = (HttpURLConnection) url.openConnection();
         connection.setRequestMethod("GET");
         int responseCode = connection.getResponseCode();
         Assert.assertEquals("HTTP 403 Forbidden expected for urls with " +
                 "directory change", HttpURLConnection.HTTP_FORBIDDEN, responseCode);
     }
+
+    private void waitForFrontendCompilation(URL url) throws IOException, InterruptedException {
+        boolean frontendCompiled = false;
+        int attemptsRemaining = 100;
+        while (!frontendCompiled && attemptsRemaining-- > 0) {
```

If for some reason the frontend compilation is not be completed in a given timeout, then the test execution would go further and fail because of unexpected response code.
Whether it's better to fail fast the test with an appropriate message when it's reach the timeout?
I would help analyse the root cause in the future.

---

It's true. I will change it to throw an exception